app/domain/authentication/authn_jwt/restriction_validation/validate_restrictions_included.rb

```diff
@@ -0,0 +1,72 @@
+module Authentication
+  module AuthnJwt
+    module RestrictionValidation
+      # This class is responsible for retrieving the correct value from the JWT token
+      # of the requested attribute.
+      class ValidateRestrictionsIncluded
+        def initialize(
+          decoded_token:,
+          aliased_claims:,
+          logger: Rails.logger
+        )
+          @decoded_token = decoded_token
+          @aliased_claims = aliased_claims
+          @logger = logger
+        end
+
+        def valid_restriction?(restriction)
+          annotation_name = restriction.name
+          claim_name = claim_name(annotation_name)
+
+          validate_annotation_value(restriction)
+
+          unless @decoded_token.key?(claim_name)
+            raise Errors::Authentication::AuthnJwt::JwtTokenClaimIsMissing,
+                  claim_name_for_error(annotation_name, claim_name)
+          end
+
+          token_values = @decoded_token.fetch(claim_name)
+
+          unless token_values.is_a?(Array)
+            raise Errors::Authentication::ResourceRestrictions::InconsistentHostAnnotationType, annotation_name
+          end
+
+          parsed_annotation_values = parse_annotation_values(restriction)
```

Comment:
I think that we should suppress multiple calls of `parse_annotation_values` especially because we cannot to memorize the result.
I guess that parse can take a validation function on itself from the main method point of view